File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"

#include <limits.h>
#include <stdlib.h>
#include <math.h>

int ability(void) {
    int sum = 0;
    int min = INT_MAX;

    for (int i=0; i<4; i++) {
        int roll = rand() % 6 + 1;
        if (roll < min) min = roll;
        sum += roll;
    }
    
    return sum - min;
}

int modifier(int score){
    return (int) floor((score - 10) / 2.);
}

dnd_character_t make_dnd_character(void) {
    dnd_character_t character;
    character.strength     = ability();
    character.dexterity    = ability();
    character.constitution = ability();
    character.intelligence = ability();
    character.wisdom       = ability();
    character.charisma     = ability();
    character.hitpoints    = 10 + modifier(character.constitution);
    return character;
}

File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
    "math/rand"
	"math"     
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return int(math.Floor(float64(score-10) / 2.0))
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	return rand.Intn(16) + 3
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	constitution := Ability()

    return Character{
        Strength: Ability(),
        Dexterity: Ability(),
        Constitution: constitution,
        Intelligence: Ability(),
        Wisdom: Ability(),
        Charisma: Ability(),
        Hitpoints: 10 + Modifier(constitution),
    }
}
